src/ngx_http_lua_kong_error_log.c

```diff
@@ -0,0 +1,183 @@
+
+
+static ngx_array_t *error_log_append_var_indexes;
+
+typedef struct {
+    ngx_str_t   name;
+    ngx_uint_t  index;
+} var_elt_t;
+
+
+static u_char *
+ngx_http_log_error_handler(ngx_http_request_t *r, ngx_http_request_t *sr,
+    u_char *buf, size_t len)
+{
+    char                      *uri_separator;
+    u_char                    *p;
+    ngx_http_upstream_t       *u;
+    ngx_http_core_srv_conf_t  *cscf;
+
+    cscf = ngx_http_get_module_srv_conf(r, ngx_http_core_module);
+
+    p = ngx_snprintf(buf, len, ", server: %V", &cscf->server_name);
+    len -= p - buf;
+    buf = p;
+
+    if (r->request_line.data == NULL && r->request_start) {
+        for (p = r->request_start; p < r->header_in->last; p++) {
+            if (*p == CR || *p == LF) {
+                break;
+            }
+        }
+
+        r->request_line.len = p - r->request_start;
+        r->request_line.data = r->request_start;
+    }
+
+    if (r->request_line.len) {
+        p = ngx_snprintf(buf, len, ", request: \"%V\"", &r->request_line);
+        len -= p - buf;
+        buf = p;
+    }
+
+    if (r != sr) {
+        p = ngx_snprintf(buf, len, ", subrequest: \"%V\"", &sr->uri);
+        len -= p - buf;
+        buf = p;
+    }
+
+    u = sr->upstream;
+
+    if (u && u->peer.name) {
+
+        uri_separator = "";
+
+#if (NGX_HAVE_UNIX_DOMAIN)
+        if (u->peer.sockaddr && u->peer.sockaddr->sa_family == AF_UNIX) {
+            uri_separator = ":";
+        }
+#endif
+
+        p = ngx_snprintf(buf, len, ", upstream: \"%V%V%s%V\"",
+                         &u->schema, u->peer.name,
+                         uri_separator, &u->uri);
+        len -= p - buf;
+        buf = p;
+    }
+
+    if (r->headers_in.host) {
+        p = ngx_snprintf(buf, len, ", host: \"%V\"",
+                         &r->headers_in.host->value);
+        len -= p - buf;
+        buf = p;
+    }
+
+    if (r->headers_in.referer) {
+        p = ngx_snprintf(buf, len, ", referrer: \"%V\"",
+                         &r->headers_in.referer->value);
+        len -= p - buf;
+        buf = p;
+    }
+
+
+    // loop through error_log_append_var_indexes and append
+    // each variable and their values to the error log
+    var_elt_t *elts  = error_log_append_var_indexes->elts;
+    ngx_uint_t nelts = error_log_append_var_indexes->nelts;
+    for (int i = 0; i < nelts; i++) {
+        var_elt_t                 *v     = &elts[i];
+        ngx_uint_t                 index = v->index;
+        ngx_http_variable_value_t *value =
+                ngx_http_get_indexed_variable(r, index);
+
+        if (value == NULL || value->not_found || !value->valid) {
+            continue;
+        }
+
+        u_char *name = v->name.data;
+        p = ngx_snprintf(buf, len, ", %s: \"%v\"", name, value);
+
+        len -= p - buf;
+        buf = p;
+    }
+
+    return buf;
+}
+
+
+static ngx_int_t
+set_error_log_handler(ngx_http_request_t *r)
+{
+    r->log_handler = ngx_http_log_error_handler;
```

Comment:
replace original r->log_handler,
note that orignal r->log_handler is initialized in https://github.com/search?q=repo%3Anginx%2Fnginx%20r-%3Elog_handler&type=code

Reply:
Yes, I suggest add some comments to describe the code and how to do in the future.

Reply:
> ❗ It replaces the nginx default logic and needs to be reviewed carefully.

correct. Agree.

> if nginx official modifies the logic [...] we need to introduce the official modification to be here

I think we could consider any minor updates to the nginx log handler optional to port here (though it would be probably a good idea for consistency) but not necessarily problematic if left out of sync. Unless of course the code used in our log handler becomes deprecated through some breaking changes in nginx.

Reply:
closing as this code no longer exists due to the patching approach (see other comments)